### PR TITLE
Automate the gathering and publishing of new toolchain versions

### LINKED ISSUE #192: Automate bumping `RUST_VERSION`
The Rust team comes out with a new version of the compiler every 6 weeks (sometimes sooner in the case of patch releases).

These releases are announced on the Rust blog as well as via GitHub Releases. Minor releases, e.g. Rust 1.75.0, always drop on Thursdays.

I wonder if there is something like an RSS feed or some magic we can run on a cron by querying the GitHub API to pull the latest released version of the compiler, update the `Rakefile`, run `bundle exec rake toolchain:sync`, commit + push + open a PR.

This is how `generate_third_party` does this: https://github.com/artichoke/generate_third_party/blob/9c76d24d2276cba48c1b90032837d6c59cdda7b9/lib/generate_third_party/cargo_about/platform.rb#L28-L50

### PR DESCRIPTION
**Pull request by b-n**
An attempt at automatically updating the toolchain version.

In short what this PR does:

- Get the current RUST_VERSION from the Rakefile
  I originally though of using some grep/regex magic to get this value, but thought: "Ruby knows best what it's looking for".
- Get the latest release from rust-lang/rust (This is what github determines as "latest" - so there is some limitation there if patches for older versions are released)
- IF the values are different:
  - update the value with sed magic in the file (Makes me second guess my "ruby knows best" comment above) and runs `bundle exec rake toolchain:sync`
  - Create[/Update] a PR to merge.

I've had some success running this action with [act](https://github.com/peter-evans/create-pull-request/issues/633), however for some reason I'm getting a "Could not read from remote repository" when it tries to run `git remote prune origin` here: https://github.com/peter-evans/create-pull-request/blob/main/src/create-pull-request.ts#L138. (running with `act -s GITHUB_TOKEN="$(gh auth token)" -j check-toolchain-version`, have also tried a hardcoded PAT token, to no success).

I can confirm everything works up until the git pr/commit part of it, so that's all good. In theory the last PR part should "Just work", but it might be a wait and see on the next rust release. I fear having a dirty tree (only easy way to test this) is causing some issue somewhere.

Fixes: #192

### CODE REVIEW

File: .github/workflows/toolchain-version-check.yaml
Comment (lopopolo) [inline .github/workflows/toolchain-version-check.yaml]: Rust releases go out every 6 weeks on Thursdays. We can probably limit this to run once a week on Fridays, maybe at 5am UTC to avoid racing with the nightly release job?

File: .github/workflows/toolchain-version-check.yaml
Comment (lopopolo) [inline .github/workflows/toolchain-version-check.yaml]: let's use a SHA to pin the version. We only use tag versions for GitHub actions in the @actions and @artichoke GitHub organizations.

File: .github/workflows/toolchain-version-check.yaml
Comment (lopopolo) [inline .github/workflows/toolchain-version-check.yaml:32]: this is super neat!

File: .github/workflows/toolchain-version-check.yaml
Comment (lopopolo) [inline .github/workflows/toolchain-version-check.yaml]: typo fix

```suggestion
          # Matches "RUST_VERSION = 'M.M.P'" and replaces with "RUST_VERSION = '$NEXT_VERSION'"
```

File: .github/workflows/toolchain-version-check.yaml
Comment (lopopolo) [inline .github/workflows/toolchain-version-check.yaml]: ditto, let's pin by SHA

File: .github/workflows/toolchain-version-check.yaml
Comment (lopopolo) [inline .github/workflows/toolchain-version-check.yaml]: typo fix

```suggestion
            ⚠ This PR is automatically created via a Github Action. Ensure all changes are valid before merging.
```

File: .github/workflows/toolchain-version-check.yaml
Comment (lopopolo) [inline .github/workflows/toolchain-version-check.yaml]: can we put something like from ${old_version} to ${new_version} in the PR text?

File: .github/workflows/toolchain-version-check.yaml
Comment (lopopolo) [inline .github/workflows/toolchain-version-check.yaml:78]: let's put the rust version in the branch name

File: .github/workflows/toolchain-version-check.yaml
Comment (lopopolo) [inline .github/workflows/toolchain-version-check.yaml:76]: I'd prefer this to be something like:

> Update Rust toolchain to ${new_version}

File: .github/workflows/toolchain-version-check.yaml
Comment (lopopolo) [inline .github/workflows/toolchain-version-check.yaml]: I'd prefer this to be something like:

> Update Rust toolchain to ${new_version}

File: .github/workflows/toolchain-version-check.yaml
Comment (lopopolo) [inline .github/workflows/toolchain-version-check.yaml]: can you use the @artichoke/contributors GitHub team?

File: .github/workflows/toolchain-version-check.yaml
Comment (lopopolo) [inline .github/workflows/toolchain-version-check.yaml:76]: Does this do the equivalent of `git add . && git commit -m '...'`?

File: .github/workflows/toolchain-version-check.yaml
Comment (lopopolo) [inline .github/workflows/toolchain-version-check.yaml]:
```suggestion
    name: Check Toolchain Version
```

**Review by lopopolo — CHANGES_REQUESTED** [review]
this looks super excellent! thanks for putting this together.

A few changes we should make before merging.

File: .github/workflows/toolchain-version-check.yaml
Comment (b-n) [reply .github/workflows/toolchain-version-check.yaml]: I was curious why some where SHA's. I assume to fix any risk of supply chain attacks and floating tag issues. I'm curious, does it work with dependabot?

File: .github/workflows/toolchain-version-check.yaml
Comment (b-n) [reply .github/workflows/toolchain-version-check.yaml:78]: One nice feature of this action is that it'll continuously update the PR if the branch name stays the same (e.g. like dependabot). If we change the name of the branch, I'd expect to see more than one PR hitting the repo (assuming it's not merged for example). Thoughts?

File: .github/workflows/toolchain-version-check.yaml
Comment (b-n) [reply .github/workflows/toolchain-version-check.yaml]: I'll have a play and see what can be done. Until now I've not actually been able to get a test working for this on this current branch, so will test on another repo and see what can be done.

File: .github/workflows/toolchain-version-check.yaml
Comment (b-n) [reply .github/workflows/toolchain-version-check.yaml:76]: As with 2 comments ago, will test on another repo and see what's achieveable.

File: .github/workflows/toolchain-version-check.yaml
Comment (b-n) [reply .github/workflows/toolchain-version-check.yaml]: ditto

File: .github/workflows/toolchain-version-check.yaml
Comment (b-n) [reply .github/workflows/toolchain-version-check.yaml:76]: From what I read on the docs, yep! https://github.com/peter-evans/create-pull-request#action-behaviour

I'll test it on a private repo for certainty though.

File: .github/workflows/toolchain-version-check.yaml
Comment (b-n) [reply .github/workflows/toolchain-version-check.yaml]: I changed the text ever so slightly :smile:

File: .github/workflows/toolchain-version-check.yaml
Comment (lopopolo) [reply .github/workflows/toolchain-version-check.yaml]: yes it does. Dependabot will also update the comment which indicates the tag.

File: .github/workflows/toolchain-version-check.yaml
Comment (lopopolo) [inline .github/workflows/toolchain-version-check.yaml]: looks like you missed the comment with the version tag on this one

File: .github/workflows/toolchain-version-check.yaml
Comment (lopopolo) [reply .github/workflows/toolchain-version-check.yaml:78]: oh gotcha. in that case, this is fine as is.

**Comment by b-n** [comment]
@lopopolo Woo! https://github.com/b-n/gh-action-test/blob/main/.github/workflows/create-pr.yaml generates https://github.com/b-n/gh-action-test/pull/1 so the answer is yes - we can play with variables, and yes - it works. I added you as a contributor to that repo so you can press the play button here: https://github.com/b-n/gh-action-test/actions/workflows/create-pr.yaml (if you just want to see it working :sweat_smile:

I'll fix up this PR tomorrow at some point. The key takeaways:
- It works
- It supports variable interpolation
- It does require some permissions on the action runner - I'll see if I can find the values in terraform and add to project-infrastructure if applicable.

**Comment by b-n** [comment]
w.r.t runner permissions: I can't find a terraform setting for this specific switch: https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#preventing-github-actions-from-creating-or-approving-pull-requests.

Apparently it's possible to just us `gh` commands on github runners without having to use this particular setting (haven't tested). It seems a bit redundant though whether it's from a action or whether it's from your own `gh` scripting to restrict this functionality, so I'm not sure what's the best course.

The create-pull-request does use just the plain github api from what it looks. I do wonder whether providing a machine token to the `token:` attribute (e.g. a machine PAT token) would actually circumvent this setting. Is the one available for the artichoke org?

File: .github/workflows/toolchain-version-check.yaml
Comment (b-n) [reply .github/workflows/toolchain-version-check.yaml]: Done and done!

File: .github/workflows/toolchain-version-check.yaml
Comment (b-n) [reply .github/workflows/toolchain-version-check.yaml:76]: I've got a from/to range, feel free to propose an alternative :smile:

Comment (b-n) [reply .github/workflows/toolchain-version-check.yaml:76]: Can confirm, this is a yes. It can be configured with specific adds and/or specific removal of certain files being committed if need be. I imagine the `toolchain:sync` will always update all "needed files", so IMO `git add .` makes sense here.

File: .github/workflows/toolchain-version-check.yaml
Comment (b-n) [inline .github/workflows/toolchain-version-check.yaml:9]: Name of file = "toolchain version check"
name of the job = "check toolchain version".

Should we unify these? Happy either way, just pointing it out.

File: .github/workflows/toolchain-version-check.yaml
Comment (b-n) [inline .github/workflows/toolchain-version-check.yaml]: A bit of icky bash magic. Maybe an easier to read alternative is to make a `$MATCHES` env var since the interpolation in the `${{ }}` is javascript friendly from memory - e.g. could make a ternary there.

Output in `act`:
```console
[Toolchain Version Check/Check toolchain version]   ✅  Success - Main Extract rust version info
[Toolchain Version Check/Check toolchain version]   ⚙  ::set-output:: next=1.75.0
[Toolchain Version Check/Check toolchain version]   ⚙  ::set-output:: matches=false
[Toolchain Version Check/Check toolchain version]   ⚙  ::set-output:: current=1.74.0
```

File: .github/workflows/toolchain-version-check.yaml
Comment (b-n) [inline .github/workflows/toolchain-version-check.yaml:75]: e.g. https://github.com/rust-lang/rust/compare/1.74.1...1.75.0

**Comment by b-n** [comment]
@lopopolo I have one tiny other little idea:

- Even though not "needed", we can create a `rust-toolchain.toml` file in this repo.
- In the rakefile, we read the current version from `rust-toolchain.toml` instead of inside the artichoke main repo: https://github.com/artichoke/artichoke/blob/trunk/Rakefile#L201
- We read/update the value in `rust-toolchain.toml` via grep/sed.

Why? Because then we can use the exact same github action in this repo as we can in the main artichoke repo. Thoughts?

Could also make it a published action. e.g.

```yaml
- uses: artichoke/rust-toolchain-dependabot
  with:
    toolchain_path: ./rust-toolchain.toml
    toolchain_syntax: toml
    pre_commit_command: "bundle exec rake toolchain:sync"
```

"A github action that will check and update the rust toolchain. You can provide your own command to update any other files prior to commiting"

**Comment by b-n** [comment]
Sorry, another message - eventually I might try and commit a dependabot-core for rust-toolchain. Looks like an old issue is people wanting it a long time ago: https://github.com/dependabot/dependabot-core/issues/1758#issuecomment-834333421

Dependabot-core usage restricts the running of pre-post commit things, but there is apparently a way to bump commits on top of dependabot PRs, so that could be done for the `toolchain:sync` part: https://github.com/dependabot/dependabot-core/issues/1758#issuecomment-834333421 (This is a big TIL for me today - kinda cool :sweat_smile:)

File: .github/workflows/toolchain-version-check.yaml
Comment (lopopolo) [reply .github/workflows/toolchain-version-check.yaml:9]: Doesn't bother me either way, up to you. My only ask is that if this workflow ends up appearing in multiple repos is that we do it the same.

File: .github/workflows/toolchain-version-check.yaml
Comment (lopopolo) [inline .github/workflows/toolchain-version-check.yaml]: can we add a note that we do this on Fridays because new Rust versions get released on Thursdays?

File: .github/workflows/toolchain-version-check.yaml
Comment (lopopolo) [reply .github/workflows/toolchain-version-check.yaml]: I think structuring it this way is a bit more readable.

```suggestion
          if [[ "$CURRENT" == "$NEXT" ]]; then
            echo "matches=true" >> "$GITHUB_OUTPUT"
          else
            echo "matches=false" >> "$GITHUB_OUTPUT"
          fi
```

File: .github/workflows/toolchain-version-check.yaml
Comment (lopopolo) [inline .github/workflows/toolchain-version-check.yaml]:
```suggestion
          title: Update Rust toolchain from ${{ steps.version_info.outputs.current }} to ${{ steps.version_info.outputs.next }}
```

File: .github/workflows/toolchain-version-check.yaml
Comment (lopopolo) [inline .github/workflows/toolchain-version-check.yaml]:
```suggestion
            > [!WARNING]
            > This PR is automatically created via a Github Action. Ensure all changes are valid before merging.
```

This will render like this:

> [!WARNING]
> This PR is automatically created via a Github Action. Ensure all changes are valid before merging.

File: .github/workflows/toolchain-version-check.yaml
Comment (lopopolo) [reply .github/workflows/toolchain-version-check.yaml:75]: this is awesome.

File: .github/workflows/toolchain-version-check.yaml
Comment (lopopolo) [inline .github/workflows/toolchain-version-check.yaml]:
```suggestion
        env:
          CURRENT: ${{ steps.rust_version.outputs.version }}
          NEXT: ${{ fromJson(steps.latest_rust_release.outputs.data).tag_name }}
```

File: .github/workflows/toolchain-version-check.yaml
Comment (lopopolo) [inline .github/workflows/toolchain-version-check.yaml]: This sed invocation looks a lot more understandable to me.

```suggestion
          sed -iE "s/^RUST_VERSION = \"$CURRENT\"$/RUST_VERSION = \"$NEXT\"/g" Rakefile
```

**Review by lopopolo — APPROVED** [review]
This looks pretty sweet, most of my comments are nits. Feel free to merge when you're ready.

**Comment by lopopolo** [comment]
> w.r.t runner permissions: I can't find a terraform setting for this specific switch: https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#preventing-github-actions-from-creating-or-approving-pull-requests.

at https://github.com/artichoke/docker-artichoke-nightly/settings/actions:

Screenshot 2024-01-21 at 4 25 54 PM

There's an org setting I can flip to enable us turning that on in this repo. Let me look to see if I can terraform this.

Screenshot 2024-01-21 at 4 27 15 PM

**Comment by lopopolo** [comment]
> @lopopolo I have one tiny other little idea:
> 
> * Even though not "needed", we can create a `rust-toolchain.toml` file in this repo.
> * In the rakefile, we read the current version from `rust-toolchain.toml` instead of inside the artichoke main repo: https://github.com/artichoke/artichoke/blob/trunk/Rakefile#L201
> * We read/update the value in `rust-toolchain.toml` via grep/sed.
> 
> Why? Because then we can use the exact same github action in this repo as we can in the main artichoke repo. Thoughts?
> 
> Could also make it a published action. e.g.
> 
> ```yaml
> - uses: artichoke/rust-toolchain-dependabot
>   with:
>     toolchain_path: ./rust-toolchain.toml
>     toolchain_syntax: toml
>     pre_commit_command: "bundle exec rake toolchain:sync"
> ```
> 
> "A github action that will check and update the rust toolchain. You can provide your own command to update any other files prior to commiting"

this sounds pretty sweet and I'd be interested to experiment here. Let me know if you want me to set up a repo for you

File: .github/workflows/toolchain-version-check.yaml
Comment (b-n) [reply .github/workflows/toolchain-version-check.yaml]: I very much agree - I often try to remove bash golf at work, so I'm not sure why I even added it here hah.

File: .github/workflows/toolchain-version-check.yaml
Comment (b-n) [reply .github/workflows/toolchain-version-check.yaml]: I recently also discovered this whilst making k8s training. A++.

File: .github/workflows/toolchain-version-check.yaml
Comment (b-n) [reply .github/workflows/toolchain-version-check.yaml]: For me too :smile: 

I adjusted to single quotes (that we currently have in the Rakefile), and included the other env var changes needed

**Comment by b-n** [comment]
@lopopolo I'll merge this one for now after incorporating the feedback. I'm sure the action will fail on the run unless those options are action perms are added.

Also, more than happy to give a generic action a go, so yes please to a repo!

**Comment by lopopolo** [comment]
Screenshot 2024-02-19 at 4 26 56 PM

I tweaked the setting and bumped the job. It worked!

- https://github.com/artichoke/docker-artichoke-nightly/pull/198

**Comment by lopopolo** [comment]
It looks like PRs created with this workflow don't trigger github status checks to run on PRs: https://github.com/peter-evans/create-pull-request/issues/48

A workaround I've found is to close and then reopen the PR.

**Pull request merged**
